internal/util/util.go

```diff
@@ -219,7 +219,61 @@ func FSize(f *os.File) int64 {
 // IsWordChar returns whether or not the string is a 'word character'
 // Word characters are defined as numbers, letters, or '_'
 func IsWordChar(r rune) bool {
-	return unicode.IsLetter(r) || unicode.IsNumber(r) || r == '_'
+	return IsAlphanumeric(r) || IsWordDelimiter(r)
+}
+
+// IsUpperWordChar returns whether or not the string is a 'upper word character'
+// Upper word characters are defined as numbers, upper-case letters, or delimiters ([A-Z0-9_])
+func IsUpperWordChar(r rune) bool {
+	return IsUpperAlphanumeric(r) || IsWordDelimiter(r)
+}
+
+// IsLowerWordChar returns whether or not the string is a 'lower word character'
+// Lower word characters are defined as numbers, upper-case letters, or delimiters ([a-z0-9_])
+func IsLowerWordChar(r rune) bool {
+	return IsLowerAlphanumeric(r) || IsWordDelimiter(r)
+}
+
+// IsWordDelimiter returns whether or not the string is a 'delimiter character'
+// Delimiter characters are defined as '_'
+func IsWordDelimiter(r rune) bool {
+	return r == '_'
+}
+
+// IsNoneWordChar returns whether or not the string is not a 'word character'
+// None word characters are defined as all characters not being alphanumerics, whitespaces or delimiters ([!A-Za-z0-9\s_])
+func IsNoneWordChar(r rune) bool {
+	return !IsAlphanumeric(r) && !IsWhitespace(r) && !IsWordDelimiter(r)
+}
+
+// IsAlphanumeric returns whether or not the string is an 'alphanumeric character'
+// Alphanumeric characters are defined as numbers or letters ([A-Za-z0-9])
+func IsAlphanumeric(r rune) bool {
+	return unicode.IsLetter(r) || unicode.IsNumber(r)
+}
+
+// IsUpperAlphanumeric returns whether or not the string is an 'upper alphanumeric character'
+// Upper alphanumeric characters are defined as numbers or upper-case letters ([A-Z0-9])
+func IsUpperAlphanumeric(r rune) bool {
+	return IsUpperLetter(r) || unicode.IsNumber(r)
+}
+
+// IsLowerAlphanumeric returns whether or not the string is an 'lower alphanumeric character'
+// Lower alphanumeric characters are defined as numbers or lower-case letters ([a-z0-9])
+func IsLowerAlphanumeric(r rune) bool {
+	return IsLowerLetter(r) || unicode.IsNumber(r)
+}
+
+// IsUpperLetter returns whether or not the string is a 'upper letter character'
+// Upper letter characters are defined as upper-case letters ([A-Z])
+func IsUpperLetter(r rune) bool {
+	return unicode.IsLetter(r) && unicode.IsUpper(r)
```

Comment:
`unicode.IsLetter(r)` check is redundant: `unicode.IsUpper()` and `unicode.IsLower()` return true for letters only.

Which means that `IsUpperLetter()` and `IsLowerLetter()`, as such, are also redundant?

Reply:
Probably, but it adds some meta information and follows the general idea of having a set of methods to classify runes.

By using the `IsUpperLetter` function it is clear that we are trying to verify the input to be a valid upper case **letter**. By using `unicode.IsUpper()` we check just for an upper case **char**, and by reading the documentation we might realize that this actually means that we are also testing for a **letter** as well.

Reply:
Ok, I admit I did not immediately realize myself that "a non-letter upper char" is not a thing (although that becomes obvious after a bit of thought).

But, the *implementation* of `IsUpperLetter()` doesn't need to do this unneeded `unicode.IsLetter()` check, right? It can just return `unicode.IsUpper()`?

Reply:
Right, not necessary. I probably leave a comment to explain that `unicode.IsUpper()` does the `unicode.IsLetter()` as well.